engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java

```diff
@@ -465,6 +470,66 @@ private Class<?> findNestedClass(Class<?> enclosingClass, String nestedClassName
         return m.get(nestedClassName);
     }
 
+    private boolean pyToJavaReplaced(final Class<?> scope, final MethodCallExpr n) {
+        final String methodName = n.getNameAsString();
+        final QueryScope queryScope = ExecutionContext.getContext().getQueryScope();
+
+        if (scope == null) {
+            final Class<?> methodClass = variables.get(methodName);
+            if (methodClass == PyCallableWrapper.class) {
+                final Object paramValueRaw = queryScope.readParamValue(methodName, null);
+                final PyCallableWrapper pyCallableWrapper = (PyCallableWrapper) paramValueRaw;
+                try {
+                    String javaMethodName = pyCallableWrapper.getAttribute("_j_simple_name", String.class);
+                    n.setName(javaMethodName);
+                    int nargs = pyCallableWrapper.getAttribute("_nargs", int.class);
+                    PyListWrapper defArgList = (PyListWrapper) pyCallableWrapper.getAttribute("_def_args").asList();
+                    int defArgsToAdd = nargs - n.getArguments().size();
+                    if (defArgsToAdd > defArgList.size()) {
+                        throw new IllegalArgumentException("Missing args for " + methodName);
+                    }
+                    if (defArgsToAdd < 0) {
+                        // do nothing, could be valid if the java method has a vararg
+                    }
+                    // Since the func call follows the Java syntax, we don't do name matching for the keyword args
+                    for (int i = defArgList.size() - (nargs - n.getArguments().size()); i < defArgList.size(); i++) {
+                        Object v = PythonScopeJpyImpl.convert(defArgList.get(i));
+                        n.addArgument(v.getClass() == String.class ? "\"" + v.toString() + "\"" : v.toString());
+                    }
+                    return true;
+                } catch (IllegalArgumentException iae) {
+                    throw iae;
+                } catch (RuntimeException e) {
+                    // Not a Java replaceable callable, can safely ignore
```

Comment:
Can we use `hasAttribute` to avoid using failed `getAttribute` calls for control flow?